scipy/sparse/tests/test_base.py

```diff
@@ -4868,6 +4907,51 @@ def test_tocompressed_duplicates(self):
         csc = coo.tocsc()
         assert_equal(csc.nnz + 2, coo.nnz)
 
+    def test_has_canonical_format(self):
+        "Ensure has_canonical_format memoizes state for sum_duplicates"
+
+        A = self.coo_container((2, 3))
+        assert_equal(A.has_canonical_format, True)
+
+        A_array = np.array([[0, 2, 0]])
+        A_coords_form = (np.array([2]), (np.array([0]), np.array([1])))
+        A_coords_dups = (np.array([1, 1]), (np.array([0, 0]), np.array([1, 1])))
+
+        A = self.coo_container(A_array)
+        assert A.has_canonical_format is True
+        A = self.coo_container(A_coords_form)
+        assert A.has_canonical_format is False
+        A.sum_duplicates()
+        assert A.has_canonical_format is True
+
+        A = self.coo_container(A, copy=True)
+        assert A.has_canonical_format is True
+        A = self.coo_container(A, copy=False)
+        assert A.has_canonical_format is False
```

Comment:
Why does the `copy` flag affect whether it the flag is True or False?

Reply:
When `copy` is True, and the input is sparse, the `has_canonical_format` flag is set based on the incoming sparse flag value. When `copy` is False the flag is set to False.  This [happens in `_coo.py`](https://github.com/scipy/scipy/blob/e27588f9f232a525eb130f7dc6243c0ef12887ac/scipy/sparse/_coo.py#L66) 
I'm not sue of the history for why not copy from the incoming flag to the new flag no matter what the `copy` argument is. I think that would be a good thing to make more consistent when we change the code.  Let's set the has_canonical_format flag for COO format based on the flag's value on the incoming sparse object.